### PR TITLE
Multi Column Feature for Gconstruct

### PR DESCRIPTION
**Pull request by jalencato**
*Issue #, if available:*

*Description of changes:*

Draft for allow multi column feature for gconstruct

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/gconstruct/transform.py [hunk 1]
```diff
@@ -1061,14 +1076,19 @@ def preprocess_features(data, ops):
...
-        for key, val in res.items():
-            pre_data[key] = val
+        if isinstance(op.col_name, str):
+            col_name = [op.col_name]
+        else:
+            col_name = op.col_name
+        for col in col_name:
+            res = op.pre_process(data[col])
+            assert isinstance(res, dict)
+            for key, val in res.items():
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:1086]: There is a bug here.

File: python/graphstorm/gconstruct/transform.py [hunk 2]
```diff
@@ -1080,23 +1100,55 @@ def process_features(data, ops):
...
+            res = op(data[col])
+            assert isinstance(res, dict)
+            for key, val in res.items():
+                # Check if it has 1D features. If yes, convert to 2D features
+                if len(val.shape) == 1:
+                    if isinstance(val, ExtMemArrayWrapper):
+                        val = val.to_numpy().reshape(-1, 1)
+                    else:
+                        val = val.reshape(-1, 1)
+                if len(col_name) == 1:
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:1131]: What if res.items() returns multiple key, val pairs? Are we going to only store the first one?

File: python/graphstorm/gconstruct/transform.py [hunk 3]
```diff
@@ -1080,23 +1100,55 @@ def process_features(data, ops):
...
-                if isinstance(val, ExtMemArrayWrapper):
-                    val = val.to_numpy().reshape(-1, 1)
+        hash_hex_feature_path = generate_hash()
+        feature_path = 'feature_{}_{}'.format(op.feat_name, hash_hex_feature_path)
+        feature_path = ext_mem + feature_path if ext_mem is not None \
+            else feature_path
+        if isinstance(op.col_name, str):
+            col_name = [op.col_name]
+        else:
+            col_name = op.col_name
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:1119]: We can move L1112 to L1114 here and change the logic into something like:
```
if ext_mem is not None:
    hash_hex_feature_path = generate_hash()
    feature_path = 'feature_{}_{}'.format(op.feat_name, hash_hex_feature_path)
    feature_path = os.path.join(ext_mem, feature_path)
    os.makedirs(feature_path)
    wrapper = ExtFeatureWrapper(feature_path
else:
    wrapper = None
```

File: python/graphstorm/gconstruct/transform.py [hunk 4]
```diff
@@ -1080,23 +1100,55 @@ def process_features(data, ops):
...
+                        val = val.reshape(-1, 1)
+                if len(col_name) == 1:
+                    new_data[key] = val
+                    continue
+                tmp_key = key
+                # Use external memory if it is available
+                if ext_mem is not None:
+                    if not os.path.exists(feature_path):
+                        os.makedirs(feature_path)
+                        wrapper = ExtFeatureWrapper(feature_path, val.shape, val.dtype)
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:1139]: Can we create ExtFeatureWrapper before L1121?

File: python/graphstorm/gconstruct/transform.py [hunk 5]
```diff
@@ -1080,23 +1100,55 @@ def process_features(data, ops):
...
+
+                    hash_hex = generate_hash()
+                    val_path = feature_path + '/{}_{}.npy'.format(col, hash_hex)
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:1142]: Why we need val_path here?

File: python/graphstorm/gconstruct/utils.py [hunk 6]
```diff
@@ -141,6 +144,14 @@ def _estimate_sizeof(data):
 
     return data_size
 
+def generate_hash():
+    """ Generate unique hashcode
+    """
+    random_uuid = uuid.uuid4()
```
Comment (classicsong) [inline python/graphstorm/gconstruct/utils.py:150]: Can we use uuid directly?

File: python/graphstorm/gconstruct/utils.py [hunk 7]
```diff
@@ -530,6 +541,117 @@ def to_tensor(self):
...
+        The shape of the array.
+    dtype : numpy dtype
+        The data type.
+    merged_file: str
+        The merged file name
+    """
+    def __init__(self, arr_path, shape, dtype,
+                 merged_file="merged_feature.npy"):
+        self._arr_path = arr_path
+        self._shape = (shape[0], 0)
```
Comment (classicsong) [inline python/graphstorm/gconstruct/utils.py:561]: Do we really need shape here?

File: python/graphstorm/gconstruct/utils.py [hunk 8]
```diff
@@ -530,6 +541,117 @@ def to_tensor(self):
...
+        self._orig_dtype = self._dtype = dtype
+        self._arr = None
+        self.merged_file = merged_file
+        self.wrapper = []
+
+    @property
```
Comment (classicsong) [inline python/graphstorm/gconstruct/utils.py:567]: Can we inherit from ExtNumpyWrapper? Many functions are same. And this class can be treated as a sub-class of ExtNumpyWrapper as it also uses Numpy mmap.

File: python/graphstorm/gconstruct/utils.py [hunk 9]
```diff
@@ -530,6 +541,117 @@ def to_tensor(self):
...
+            return arr
+        else:
+            return self._arr.astype(self._dtype)
+
+    def to_tensor(self):
+        """ Return Pytorch tensor.
+        """
+        return th.tensor(self.to_numpy())
+
+    def append(self, wrap, path=None):
```
Comment (classicsong) [inline python/graphstorm/gconstruct/utils.py:625]: The path is not necessary, we can generate a temporal path.
BTW, please add doc strings.

File: python/graphstorm/gconstruct/utils.py [hunk 10]
```diff
@@ -530,6 +541,117 @@ def to_tensor(self):
...
+            ext_val[:] = wrap[:]
+            ext_val.flush()
+            wrap = ExtNumpyWrapper(path, wrap.shape, wrap.dtype)
+        self.wrapper.append(wrap)
+
+    def merge(self):
+        """ Return feature col-wised.
+        """
+        for wrap in self.wrapper:
+            self._shape = (wrap.shape[0], self._shape[1] + wrap.shape[1])
```
Comment (classicsong) [inline python/graphstorm/gconstruct/utils.py:643]: Can you do something like:
```
feat_dim = sum([wrap.shape[1] for wrap in self.wrapper])
self._shape = (wrap.shape[0], feat_dim)
```

File: python/graphstorm/gconstruct/transform.py (hunk 5 above)
Comment (jalencato) [reply python/graphstorm/gconstruct/transform.py:1142]: I think we can put the val_path inside the function.

File: python/graphstorm/gconstruct/utils.py (hunk 6 above)
Comment (jalencato) [reply python/graphstorm/gconstruct/utils.py:150]: Let me check if we can directly use str(random_uuid), uuid here is an object, so we need to use the string conversion

File: python/graphstorm/gconstruct/utils.py (hunk 8 above)
Comment (jalencato) [reply python/graphstorm/gconstruct/utils.py:567]: The only concern I did not inherit is about the shape definition here. Let me check if we can do so. Mainly about the shape.

File: python/graphstorm/gconstruct/utils.py (hunk 9 above)
Comment (jalencato) [reply python/graphstorm/gconstruct/utils.py:625]: Let me move the outside path inside so we will not need temporal path here.

File: python/graphstorm/gconstruct/transform.py (hunk 2 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:1142]: You did not resolve my question.

Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:1142]: If we only expect res to return 1 (key, val). Please add an assert here and also a TODO.

File: python/graphstorm/gconstruct/utils.py [hunk 11]
```diff
@@ -530,6 +538,91 @@ def to_tensor(self):
         """
         return th.tensor(self.to_numpy())
 
+class ExtFeatureWrapper(ExtNumpyWrapper):
+    """ The wrapper to memory-mapped Numpy array when combining feature
+
+    Parameters
+    ----------
+    arr_path : str
+        The path to different feature files directory
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/gconstruct/utils.py:547]: and dtype

File: python/graphstorm/gconstruct/utils.py [hunk 12]
```diff
@@ -530,6 +538,91 @@ def to_tensor(self):
...
+
+    def __del__(self):
+        self.cleanup()
+
+    def cleanup(self):
+        """ Clean up the array.
+        """
+        self._arr.flush()
+        self._arr = None
+        shutil.rmtree(self.directory_path)
```
Comment (classicsong) [inline python/graphstorm/gconstruct/utils.py:576]: Why remove the entire directory, not the self._arr_path?

File: python/graphstorm/gconstruct/utils.py [hunk 13]
```diff
@@ -530,6 +538,91 @@ def to_tensor(self):
...
+            return self._arr.astype(self._dtype)
+
+    def append(self, wrap):
+        """Add external memory wrapper to the wrapper
+        """
+        if not self.shape:
+            self._shape = (wrap.shape[0], 0)
+        if self.shape and self.shape[0] != wrap.shape[0]:
+            raise RuntimeError("Expect that ExtFeatureWrapper has a "
+                               "first dimension that is the same")
```
Comment (classicsong) [inline python/graphstorm/gconstruct/utils.py:598]: Expect that the first dimension of the ExtFeatureWrapper and wrap is same but get xxx and xxx.

File: python/graphstorm/gconstruct/utils.py [hunk 14]
```diff
@@ -530,6 +538,91 @@ def to_tensor(self):
...
+            raise RuntimeError("The ExtFeatureWrapper needs to be merge first")
+        if self._arr is None:
+            arr = np.memmap(self._arr_path, self._orig_dtype, mode="r", shape=self._shape)
+            if self._dtype != self._orig_dtype:
+                arr = arr.astype(self._dtype)
+            return arr
+        else:
+            return self._arr.astype(self._dtype)
+
+    def append(self, wrap):
```
Comment (classicsong) [inline python/graphstorm/gconstruct/utils.py:591]: Add docstring for parameters.

File: python/graphstorm/gconstruct/utils.py [hunk 15]
```diff
@@ -530,6 +538,91 @@ def to_tensor(self):
...
+        """Add external memory wrapper to the wrapper
```
Comment (classicsong) [inline python/graphstorm/gconstruct/utils.py:592]:
```suggestion
        """Add an external memory wrapper.
```

Comment (classicsong) [reply python/graphstorm/gconstruct/utils.py:592]: Actually, the input may not be an wrapper, it can also be an array.
You can just call it feature.

File: python/graphstorm/gconstruct/utils.py [hunk 16]
```diff
@@ -530,6 +538,91 @@ def to_tensor(self):
...
+            ext_val = np.memmap(path, wrap.dtype, mode="w+", shape=wrap.shape)
+            ext_val[:] = wrap[:]
+            ext_val.flush()
+            wrap = ExtNumpyWrapper(path, wrap.shape, wrap.dtype)
+        self.wrapper.append(wrap)
+
+    def merge(self):
+        """ Merge feature col-wised.
+        """
+        assert self.wrapper, "Cannot merge None list, " \
```
Comment (classicsong) [inline python/graphstorm/gconstruct/utils.py:611]:
```suggestion
        assert self.wrapper, "Cannot merge an empty list, " \
```

File: tests/unit-tests/gconstruct/test_construct_graph.py [hunk 17]
```diff
@@ -1274,6 +1277,407 @@ def test_parse_edge_data():
...
+        assert str(e) == "Do not support categorical feature " \
+                         "transformation on multiple columns"
+
+    feat_op8 = [{
+        "feature_col": [],
+        "feature_name": "test3"
+    }]
+    try:
+        (res, _, _) = parse_feat_ops(feat_op8)
+        assert False, "expected Error raised for invalid feature column"
```
Comment (classicsong) [inline tests/unit-tests/gconstruct/test_construct_graph.py:1560]: Please be more specific of what corner case you are testing. So other maintainer can understand what error you are testing here.

File: python/graphstorm/gconstruct/transform.py [hunk 18]
```diff
@@ -1061,14 +1082,29 @@ def preprocess_features(data, ops):
...
-        assert isinstance(res, dict)
-        for key, val in res.items():
-            pre_data[key] = val
+        if isinstance(op.col_name, str):
+            col_name = [op.col_name]
+        else:
+            col_name = op.col_name
+        for col in col_name:
+            res = op.pre_process(data[col])
+            # Do not expect multiple keys for multiple columns, expect output is
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:1090]:
```suggestion
            # Do not expect multiple keys for multiple columns, the expected output will only have 1 key/val pair.
```

File: python/graphstorm/gconstruct/transform.py [hunk 19]
```diff
@@ -1061,14 +1082,29 @@ def preprocess_features(data, ops):
...
+            # {feat_name: encoding}. But for single column, some feature transformation like
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:1091]:
```suggestion
            # But for single column, some feature transformations like
```

File: python/graphstorm/gconstruct/transform.py [hunk 20]
```diff
@@ -1080,23 +1116,60 @@ def process_features(data, ops):
...
+                if len(val.shape) == 1:
+                    if isinstance(val, ExtMemArrayWrapper):
+                        val = val.to_numpy().reshape(-1, 1)
+                    else:
+                        val = val.reshape(-1, 1)
+                if len(col_name) == 1:
+                    new_data[key] = val
+                    continue
+                tmp_key = key
+                # Use external memory if it is available
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:1161]:
```suggestion
                # Use external memory if it is required
```

File: python/graphstorm/gconstruct/transform.py [hunk 21]
```diff
@@ -1080,23 +1116,60 @@ def process_features(data, ops):
...
+            hash_hex_feature_path = generate_hash()
+            feature_path = 'feature_{}_{}'.format(op.feat_name, hash_hex_feature_path)
+            feature_path = ext_mem_path + feature_path
+            os.makedirs(feature_path)
+            wrapper = ExtFeatureWrapper(feature_path)
+        else:
+            wrapper = None
+        for col in col_name:
+            res = op(data[col])
+            # Do not expect multiple keys for multiple columns, expect output is
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:1143]: Same comment as in preprocess_features.

File: python/graphstorm/gconstruct/transform.py [hunk 22]
```diff
@@ -1080,23 +1116,60 @@ def process_features(data, ops):
...
+            # {feat_name: encoding}. But for single column, some feature transformation like
+            # Tokenizer will return multiple key-val pairs, so do not check for single column
+            if len(col_name) > 1:
+                assert isinstance(res, dict) and len(res) == 1, \
+                    f"It is expected only have one feature name after the process_features " \
+                    f"for multiple columns, but get {len(res)}"
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:1149]:
```suggestion
                    f"for multiple column feature transformation, but get {len(res)}"
```

File: python/graphstorm/gconstruct/utils.py [hunk 23]
```diff
@@ -530,6 +538,107 @@ def to_tensor(self):
         """
         return th.tensor(self.to_numpy())
 
+class ExtFeatureWrapper(ExtNumpyWrapper):
+    """ The wrapper to memory-mapped Numpy array when combining features
+
+    Parameters
+    ----------
+    arr_path : str
+        The path to different feature files directory
```
Comment (classicsong) [inline python/graphstorm/gconstruct/utils.py:547]:
```suggestion
        A path to the directory of different feature files.
```

File: python/graphstorm/gconstruct/utils.py [hunk 24]
```diff
@@ -530,6 +538,107 @@ def to_tensor(self):
...
+            The value needs to be packed.
+        """
+        if not self._shape:
+            self._shape = (feature.shape[0], 0)
+        if self._shape and self._shape[0] != feature.shape[0]:
+            raise RuntimeError(f"Expect that ExtFeatureWrapper has a "
+                               f"first dimension that is the same but get "
+                               f"{self.shape[0]} and {feature.shape[0]}")
+        # Convert the numpy array into an ExtNumpyWrapper
+        # By converting into an ExtNumpyWrapper, it is possible to avoid loading
```
Comment (classicsong) [inline python/graphstorm/gconstruct/utils.py:611]:
```suggestion
        # By converting it into an ExtNumpyWrapper, it will avoid loading
```

File: python/graphstorm/gconstruct/utils.py [hunk 25]
```diff
@@ -530,6 +538,107 @@ def to_tensor(self):
...
+    def __getitem__(self, idx):
+        if not self._shape:
+            raise RuntimeError("Call ExtFeatureWrapper.merge() first before calling __getitem__")
+        if self._arr is None:
+            self._arr = np.memmap(self._arr_path, self._dtype, mode="r",
+                                  shape=self._shape)
+        return self._arr[idx].astype(self.dtype)
+
+    def __del__(self):
+        self.cleanup()
```
Comment (classicsong) [inline python/graphstorm/gconstruct/utils.py:571]: Can you add a unitest to check if __del__ works in the unit test?

File: python/graphstorm/gconstruct/transform.py [hunk 26]
```diff
@@ -989,6 +995,13 @@ def parse_feat_ops(confs):
                                       infer_batch_size=infer_batch_size,
                                       out_dtype=out_dtype)
             elif conf['name'] == 'max_min_norm':
+                # TODO: Not support max_min_norm feature transformation on multiple columns
+                # without explicitly defining max_val and min_val. Otherwise, it is not clear to
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:999]:
```suggestion
                # without explicitly defining max_val and min_val. Otherwise, the definition of max_val and min_val for each column is unclear.
```

File: python/graphstorm/gconstruct/utils.py (hunk 25 above)
Comment (jalencato) [reply python/graphstorm/gconstruct/utils.py:571]: It is already there, checkout https://github.com/awslabs/graphstorm/blob/7d140e150d1fc3c3ef12fbaa9b55e17d0915b429/tests/unit-tests/gconstruct/test_construct_graph.py#L1686.

**Review by classicsong — APPROVED** [review]
LGTM
